command/agent/agent.go

```diff
@@ -600,6 +608,67 @@ func (a *Agent) setupClient() error {
 	return nil
 }
 
+// setupNodeID will pull the persisted node ID, if any, or create a random one
+// and persist it.
+func (a *Agent) setupNodeID(config *Config) error {
+	// If they've configured a node ID manually then just use that, as
+	// long as it's valid.
+	if config.NodeID != "" {
+		if _, err := uuid.ParseUUID(string(config.NodeID)); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// For dev mode we have no filesystem access so just make a GUID.
+	if a.config.DevMode {
+		id, err := uuid.GenerateUUID()
+		if err != nil {
+			return err
+		}
+
+		config.NodeID = types.NodeID(id)
+		a.logger.Printf("[INFO] agent: Generated unique node ID %q for this agent (will not be persisted in dev mode)", config.NodeID)
+		return nil
+	}
+
+	// Load saved state, if any. Since a user could edit this, we also
+	// validate it.
+	fileID := filepath.Join(config.DataDir, "node-id")
+	if _, err := os.Stat(fileID); err == nil {
+		rawID, err := ioutil.ReadFile(fileID)
+		if err != nil {
+			return err
+		}
+
+		nodeID := strings.TrimSpace(string(rawID))
+		if _, err := uuid.ParseUUID(nodeID); err != nil {
+			return err
+		}
+
+		config.NodeID = types.NodeID(nodeID)
+	}
+
+	// If we still don't have a valid node ID, make one.
+	if config.NodeID == "" {
+		id, err := uuid.GenerateUUID()
```

Comment:
Yeah - it might not be a bad idea to save it even from `gopsutil` in case we improve the algorithm over there and people upgrade Consul in-place, but that's probably not a huge issue either way.